lib/mobility-core/src/Kernel/Utils/SlidingWindowCounters.hs

```diff
@@ -245,26 +245,91 @@ decrementByValueImpl mbTimeStamp val keyModifier key SlidingWindowOptions {..} =
 --    ** counts = {<TIMEBASED_KEY_FOR_POSITIVE_CASE>: positiveCases , <TIMEBASED_KEY_FOR_THE_TOTAL_CASES>: totalCases}
 
 -- | getLatestRatio :: (id to getResult for, and generate TIMEBASED_KEY_FOR_THE_TOTAL_CASES) -> (id modifier to create TIMEBASED_KEY_FOR_POSITIVE_CASE) -> Resultsant Ratio of the sliding window
-getLatestRatio ::
+-- Minutes | Hours | Days | Months | Years
+calculateRatioForFirstTime ::
   ( L.MonadFlow m,
     Redis.HedisFlow m r
   ) =>
   Text ->
   (Text -> Text) ->
   (Text -> Text) ->
+  (Text -> Text) ->
   SlidingWindowOptions ->
   m Double
-getLatestRatio driverId mkPostiveCaseKeyfn mkTotalCaseKeyfn s@SlidingWindowOptions {..} = do
+calculateRatioForFirstTime driverId mkPostiveCaseKeyfn mkTotalCaseKeyfn mkRatioKeyfn s@SlidingWindowOptions {..} = do
   utcTime <- L.runIO getCurrentTime
   let positiveCaseKeysList = getkeysForLastPeriods s utcTime $ makeSlidingWindowKey periodType (mkPostiveCaseKeyfn driverId)
-  let totalCountKeysList = getkeysForLastPeriods s utcTime $ makeSlidingWindowKey periodType (mkTotalCaseKeyfn driverId)
+      totalCountKeysList = getkeysForLastPeriods s utcTime $ makeSlidingWindowKey periodType (mkTotalCaseKeyfn driverId)
   positiveCases <- sum <$> mapMaybeM Redis.get positiveCaseKeysList
   totalCases <- nonZero . sum <$> mapMaybeM Redis.get totalCountKeysList
-  pure $ positiveCases / totalCases
+  let ratio = positiveCases / totalCases
+      expTime = fromInteger . floor $ diffUTCTime (incrementPeriod periodType utcTime) utcTime
+  Redis.setExp (mkTotalCaseKeyfn driverId) totalCases expTime
+  Redis.setExp (mkRatioKeyfn driverId) ratio expTime
+  pure ratio
   where
     nonZero :: Double -> Double
     nonZero a = if a == 0.0 then 1.0 else a
 
+recalculateRatio ::
+  ( L.MonadFlow m,
+    Redis.HedisFlow m r
+  ) =>
+  Text ->
+  (Text -> Text) ->
+  (Text -> Text) ->
+  (Text -> Text) ->
+  Integer ->
+  Integer ->
+  SlidingWindowOptions ->
+  m Double
+recalculateRatio driverId mkPostiveCaseKeyfn mkTotalCaseKeyfn mkRatioKeyfn changeInTotalCase changeInPositiveCase s@SlidingWindowOptions {..} = do
+  utcTime <- L.runIO getCurrentTime
+  let positiveCaseKey = mkPostiveCaseKeyfn driverId
+      totalCaseKey = mkTotalCaseKeyfn driverId
+      ratioKey = mkRatioKeyfn driverId
+  totalCase <-
+    Redis.get totalCaseKey >>= \case
+      Nothing -> findTotalCase driverId mkTotalCaseKeyfn
+      Just total -> return total
+  ratio <-
+    Redis.get ratioKey >>= \case
+      Nothing -> calculateRatioForFirstTime driverId mkPostiveCaseKeyfn mkTotalCaseKeyfn mkRatioKeyfn s
+      Just currRatio -> return currRatio
+  let positiveCase = totalCase * ratio
+      newRatio = (positiveCase + fromIntegral changeInPositiveCase) / (totalCase + fromIntegral changeInTotalCase)
+  _ <- if changeInTotalCase > 0 then Redis.incrby totalCaseKey changeInTotalCase else if changeInTotalCase < 0 then Redis.decrby totalCaseKey changeInTotalCase else pure 0
```

Comment:
can we do this part of updating total count inside increment decrement functions itself instead of doing here ?